src/tensorwaves/optimizer/callbacks.py

```diff
@@ -150,48 +122,63 @@ def __init__(
-        self.__function_call += 1
-        if self.__function_call % self.__step_size != 0:
+    def on_iteration_end(
+        self, function_call: int, logs: Optional[Dict[str, Any]] = None
+    ) -> None:
+        if logs is None:
+            return
+        if function_call % self.__step_size != 0:
             return
+        parameters = logs["parameters"]
         for par_name, value in parameters.items():
-            tf.summary.scalar(par_name, value, step=self.__function_call)
-        tf.summary.scalar(
-            "estimator", estimator_value, step=self.__function_call
-        )
+            tf.summary.scalar(par_name, value, step=function_call)
+        estimator_value = logs.get("estimator", {}).get("value", None)
+        if estimator_value is not None:
+            tf.summary.scalar("estimator", estimator_value, step=function_call)
         self.__file_writer.flush()
 
-    def finalize(self) -> None:
+    def on_function_call_end(self) -> None:
         self.__file_writer.close()
 
 
-class CallbackList(Callback):
-    """Class for combining `Callback` s.
+class YAMLSummary(Callback, Loadable):
+    def __init__(self, filename: str, step_size: int = 10) -> None:
+        """Log fit parameters and the estimator value to a `tf.summary`.
 
-    Combine different `Callback` classes in to a chain as follows:
+        The logs can be viewed with `TensorBoard
+        <https://www.tensorflow.org/tensorboard>`_ via:
 
-    >>> from tensorwaves.optimizer.callbacks import (
-    ...     CallbackList, ProgressBar, TFSummary
-    ... )
-    >>> from tensorwaves.optimizer.minuit import Minuit2
-    >>> optimizer = Minuit2(callback=CallbackList([ProgressBar(), TFSummary()]))
-    """
+        .. code-block:: bash
 
-    def __init__(self, callbacks: Iterable[Callback]) -> None:
-        self.__callbacks: List[Callback] = list()
-        for callback in callbacks:
-            self.__callbacks.append(callback)
+            tensorboard --logdir logs
+        """
+        self.__step_size = step_size
+        self.__stream = open(filename, "w")
 
-    def __call__(self, parameters: dict, estimator_value: float) -> None:
-        for callback in self.__callbacks:
-            callback(parameters, estimator_value)
+    def on_iteration_end(
+        self, function_call: int, logs: Optional[Dict[str, Any]] = None
+    ) -> None:
+        if function_call % self.__step_size != 0:
+            return
+        _empty_file(self.__stream)
+        yaml.dump(
+            logs,
+            self.__stream,
+            sort_keys=False,
+            Dumper=_IncreasedIndent,
+            default_flow_style=False,
+        )
 
-    def finalize(self) -> None:
-        for callback in self.__callbacks:
-            callback.finalize()
+    def on_function_call_end(self) -> None:
+        self.__stream.close()
+
+    @staticmethod
+    def load_latest_parameters(filename: str) -> dict:
```

Comment:
With the different interface suggestion, the loading of the state from the resource would be handled by some static function like:

```python3

@staticmethod
def load_from_file(filename:str) -> "YamlSummary":
...
```
while the usage would be
```python3
summary = YamlSummary.load_from_file("asdf.yml")
params = summary.latest_parameters()
```

Reply:
I tried to implement your suggestions, but indeed the mixing of read and write mode messes up the tests and I would have to add even more logics. Suggests bad design. (Not saying the current design is good either, but at least avoids this need for additional logic in the implementation.